atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TokenRangeWritesLogger.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableRangeMap;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.logsafe.SafeArg;
+
+public class TokenRangeWritesLogger {
+    private static final Logger log = LoggerFactory.getLogger(TokenRangeWritesLogger.class);
+
+    @VisibleForTesting
+    static final long THRESHOLD_WRITES_PER_TABLE = 100_000L;
+
+    @VisibleForTesting
+    volatile RangeMap<LightweightOppToken, ConcurrentMap<TableReference, Long>> tokenRangeWritesPerTable =
+            ImmutableRangeMap.of();
+    @VisibleForTesting
+    ConcurrentMap<TableReference, Long> writesPerTable;
+
+    public TokenRangeWritesLogger(Set<Range<LightweightOppToken>> ranges) {
```

Comment:
The issue with that is that we refreshTokenRanges every 2 minutes. Even if we only reset our counts when the list of nodes actually changes, we are still looking at resetting every time a cassandra node is down. 

For determining hotpostting it isn't really necessary to always be up to date with the token ranges, and I would rather be using slightly stale ranges than never get enough data. The case where we add new nodes is not really something I expect to happen that often, and it would be picked up the next time we create the TransactionManager. What we could do is try to refresh at fixed intervals (every 24 hours maybe?).

Reply:
We check for a change this often, but I doubt the token ranges change that often unless I misunderstand how things are allocated by Cassandra.  @ashrayjain for clarification

Reply:
Yea, the token ranges don't change that often (node being turned off doesn't change the ranges). Lacking context here but I would probably design this with callback model where the refresh token ranges can actually do something special when there is an actual change, while still being called frequently enough (ie every 2 mins).